internal/auth/oidc/service.go

```diff
@@ -2,88 +2,229 @@ package oidc
+	// OidcRepoFactory is used by "service functions" to create a new oidc repo
 	OidcRepoFactory func() (*Repository, error)
 
-	// IamRepoFactory creates a new iam repo
+	// IamRepoFactory is used by "service functions" to create a new iam repo
 	IamRepoFactory func() (*iam.Repository, error)
 
-	// AuthTokenRepFactory creates a new auth token repo
+	// AuthTokenRepFactory is used by "service functions" to create a new auth token repo
 	AuthTokenRepFactory func() (*authtoken.Repository, error)
 )
 
-// ClientInfo contains client info provide during oidc operations.
-type ClientInfo struct {
-	Type         string
-	Version      string
-	RoundTripKVs map[string]string
+// validator defines an optional interface that proto messages can implement
+// which will allow encryptMessage to validate them before encryption.
+type validator interface {
+	Validate() error
 }
 
-// StartAuth accepts a request to start an OIDC authentication/authorization
-// attempt.  It returns two URLs.  authUrl is an OIDC authorization request URL.
-// The authUrl includes a "state" parameter which is encrypted and has a payload
-// which includes (among other things) the final redirect (calculated from the
-// clientInfo), a token_request_id,  and nonce. The tokenUrl is the URL the
-// client can use to retrieve the results of the user's OIDC authentication
-// attempt. The tokenUrl contains a token_request_id, which is encrypted. No
-// options are currently supported.
-func StartAuth(ctx context.Context, kms *kms.Kms, authMethodId string, clientInfo ClientInfo) (authUrl *url.URL, tokenUrl *url.URL, e error) {
-	panic("to-do")
+// encryptMessage will encrypt the message.  The encrypted message will be wrapped in a
+// request.Wrapper and then encoded into the returned string. This function
+// supports encrypting: request.State and request.Token messages.  proto Messages should
+// implement the validator interface, so they can be validated before encryption.
+func encryptMessage(ctx context.Context, wrapper wrapping.Wrapper, am *AuthMethod, m proto.Message) (encodedEncryptedState string, e error) {
+	const op = "oidc.encryptMessage"
+	if wrapper == nil {
+		return "", errors.New(errors.InvalidParameter, op, "missing wrapper")
+	}
+	if am == nil {
+		return "", errors.New(errors.InvalidParameter, op, "missing auth method")
+	}
+	if m == nil {
+		return "", errors.New(errors.InvalidParameter, op, "missing message to encrypt")
+	}
+	switch v := m.(type) {
+	case *request.State, *request.Token:
+	default:
+		return "", errors.New(errors.InvalidParameter, op, fmt.Sprintf("unsupported message type %v for encryption", v))
+	}
+
+	v, ok := m.(validator)
+	if ok {
+		if err := v.Validate(); err != nil {
+			return "", errors.Wrap(err, op)
+		}
+	}
+	marshaled, err := proto.Marshal(m)
+	if err != nil {
+		return "", errors.Wrap(err, op, errors.WithMsg("unable to marshal message"), errors.WithCode(errors.Encode))
+	}
+	blobInfo, err := wrapper.Encrypt(ctx, marshaled, nil)
```

Comment:
It wouldn't be a bad idea to add some AAD here. Some combo of scope ID/public ID might be a good choice.

Reply:
ty!